app/controllers/BarsController.scala

```diff
@@ -78,87 +78,99 @@ class BarsController @Inject()(
-            })
-
-            for {
-              metadata <- connector.metadata(input.input.account.sortCode)
-              assess <- if (input.input.account.accountNumber.isDefined && metadata.isDefined) {
-                input.input.account.accountType match {
-                  case Some("personal") => connector.assessPersonal(
-                    input.input.subject.name.getOrElse("N/A"),
-                    input.input.account.sortCode,
-                    input.input.account.accountNumber.get, None,
-                    "bank-account-reputation-frontend").map(Some(_))
-                  case _ => connector.assessBusiness(
-                    input.input.subject.name.getOrElse("N/A"),
-                    input.input.account.sortCode,
-                    input.input.account.accountNumber.get, None,
-                    "bank-account-reputation-frontend").map(Some(_))
-                }
-              }
-              else {
-                Future.successful(None)
-              }
-            }
-            yield {
-              val dataEvent = DataEvent(
-                appConfig.appName,
-                "BankAccountReputationFrontendLookup",
-                detail = Map[String, String](
-                  "PID" -> pid.getOrElse(""),
-                  "DeviceId" -> hc.deviceID.getOrElse(""),
-                  "SortCode" -> input.input.account.sortCode,
-                  "AccountNumber" -> input.input.account.accountNumber.getOrElse(""),
-                  "AccountName" -> input.input.subject.name.getOrElse("N/A"),
-                  "AccountType" -> input.input.account.accountType.getOrElse("business")
-                ) ++ AuditDetail.from("Retrievals", retrievals)
-                  ++ AuditDetail.from("Response.metadata", metadata)
-                  ++ AuditDetail.from("Response.assess", assess))
-
-              auditConnector.sendEvent(dataEvent)
-
-              (metadata, assess) match {
-                case (None, None) => BadRequest(verifyView(inputForm.fill(input).withError("input.account.sortCode", "bars.label.sortCodeNotFound")))
-                case (Some(m), None) => Ok(verifyResultView(input.input.account, m, None))
-                case (Some(m), Some(Failure(_))) => Ok(verifyResultView(input.input.account, m, Some(BarsAssessErrorResponse())))
-                case (Some(m), Some(Success(a))) => Ok(verifyResultView(input.input.account, m, Some(a)))
-              }
+        handlePost(true, retrievals)
+      }
+  }
+
+  def handlePost(secure: Boolean = false, retrievals: Option[RetrievalsToAudit] = None)(implicit req: MessagesRequest[AnyContent]) = {
+    inputForm.bindFromRequest.fold(
+      formWithErrors => {
+        Future.successful(BadRequest(verifyView(formWithErrors, secure)))
+      },
+      input => {
+        val pid = retrievals.flatMap(r => r.credentials.collect {
+          case Credentials(pid, "PrivilegedApplication") => pid
```

Comment:
I recommend you also obtain via a retrieval the 'name' and 'credentialRole' and pass all 3 on the POST to the BE as these will end up in audit, a cheap way to see who's doing what. 

You can also log in the BE these 3 fields if the BE stores anything in Mongo.

Also, Modestatas's ActionRefiner works on each page turn, and enforces auth plus grabs the pid too.  See slack chat on this.  Modestas can chat through how it works and how to integrate this class.